File: Project2/coordinator.py
import socket
import threading
import sys
import time
import json
from datetime import datetime

class Coordinator:
    def __init__(self, port, persistence_time):
        self.port = port
        self.persistence_time = persistence_time  # in seconds
        self.members = {}  # {participant_id: {'ip': ip, 'port': port, 'status': 'online'/'offline'}}
        self.messages = []  # [{sender_id, message, timestamp}]
        self.lock = threading.Lock()
        
    def start(self):
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind(('', self.port))
        server_socket.listen(10)
        
        print(f"Coordinator started on port {self.port}")
        print(f"Persistence time threshold: {self.persistence_time} seconds")
        
        try:
            while True:
                client_socket, address = server_socket.accept()
                client_handler = threading.Thread(target=self.handle_client, args=(client_socket, address))
                client_handler.daemon = True
                client_handler.start()
        except KeyboardInterrupt:
            print("Coordinator shutting down...")
        finally:
            server_socket.close()
    
    def handle_client(self, client_socket, address):
        try:
            data = client_socket.recv(4096).decode('utf-8')
            if data:
                # Parse request
                request = json.loads(data)
                
                # Acknowledge receipt
                ack = {'status': 'ack'}
                client_socket.sendall(json.dumps(ack).encode('utf-8'))
                
                # Process request
                self.process_request(request, address)
        except Exception as e:
            print(f"Error handling client: {e}")
        finally:
            client_socket.close()
    
    def process_request(self, request, address):
        command = request.get('command')
        participant_id = request.get('participant_id')
        
        if command == 'register':
            self.register_participant(participant_id, request.get('ip'), request.get('port'))
        elif command == 'deregister':
            self.deregister_participant(participant_id)
        elif command == 'disconnect':
            self.disconnect_participant(participant_id)
        elif command == 'reconnect':
            self.reconnect_participant(participant_id, request.get('ip'), request.get('port'))
        elif command == 'msend':
            self.handle_multicast(participant_id, request.get('message'))
    
    def register_participant(self, participant_id, ip, port):
        with self.lock:
            self.members[participant_id] = {
                'ip': ip,
                'port': port,
                'status': 'online',
                'last_disconnect_time': None
            }
            print(f"Participant {participant_id} registered from {ip}:{port}")
    
    def deregister_participant(self, participant_id):
        with self.lock:
            if participant_id in self.members:
                del self.members[participant_id]
                print(f"Participant {participant_id} deregistered")
    
    def disconnect_participant(self, participant_id):
        with self.lock:
            if participant_id in self.members:
                self.members[participant_id]['status'] = 'offline'
                self.members[participant_id]['last_disconnect_time'] = time.time()
                print(f"Participant {participant_id} disconnected")
    
    def reconnect_participant(self, participant_id, ip, port):
        with self.lock:
            if participant_id in self.members:
                self.members[participant_id]['ip'] = ip
                self.members[participant_id]['port'] = port
                self.members[participant_id]['status'] = 'online'
                
                # Send missed messages within persistence time threshold
                disconnect_time = self.members[participant_id]['last_disconnect_time']
                current_time = time.time()
                
                # If participant was disconnected for less than persistence_time
                # or if disconnect_time is None (first connection), send relevant messages
                if disconnect_time is not None and current_time - disconnect_time <= self.persistence_time:
                    self.send_missed_messages(participant_id)
                
                print(f"Participant {participant_id} reconnected from {ip}:{port}")
    
    def handle_multicast(self, sender_id, message):
        current_time = time.time()
        
        # Store message for persistence
        with self.lock:
            message_entry = {
                'sender_id': sender_id,
                'message': message,
                'timestamp': current_time
            }
            self.messages.append(message_entry)
            
            # Send message to all online participants
            for participant_id, participant_info in self.members.items():
                if participant_info['status'] == 'online':
                    self.send_message_to_participant(participant_id, message_entry)
            
            print(f"Multicast message from {sender_id}: {message}")
    
    def send_missed_messages(self, participant_id):
        if participant_id not in self.members:
            return
        
        disconnect_time = self.members[participant_id]['last_disconnect_time']
        current_time = time.time()
        cutoff_time = current_time - self.persistence_time
        
        # Use the later of disconnect_time or cutoff_time
        effective_cutoff = max(disconnect_time, cutoff_time) if disconnect_time is not None else cutoff_time
        
        for message_entry in self.messages:
            # Send only messages that were sent after the participant disconnected
            # and within the persistence time threshold
            if (message_entry['timestamp'] >= effective_cutoff and 
                (disconnect_time is None or message_entry['timestamp'] >= disconnect_time)):
                self.send_message_to_participant(participant_id, message_entry)
    
    def send_message_to_participant(self, participant_id, message_entry):
        if participant_id not in self.members:
            return
            
        participant_info = self.members[participant_id]
        if participant_info['status'] != 'online':
            return
            
        try:
            client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            client_socket.connect((participant_info['ip'], participant_info['port']))
            
            payload = {
                'type': 'multicast',
                'sender_id': message_entry['sender_id'],
                'message': message_entry['message'],
                'timestamp': message_entry['timestamp']
            }
            
            client_socket.sendall(json.dumps(payload).encode('utf-8'))
            client_socket.close()
        except Exception as e:
            print(f"Error sending message to participant {participant_id}: {e}")
            # Mark participant as offline since we couldn't reach them
            with self.lock:
                self.members[participant_id]['status'] = 'offline'
                self.members[participant_id]['last_disconnect_time'] = time.time()

def main():
    if len(sys.argv) != 2:
        print("Usage: python coordinator.py <config_file>")
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    try:
        with open(config_file, 'r') as f:
            port = int(f.readline().strip())
            persistence_time = int(f.readline().strip())
    except Exception as e:
        print(f"Error reading config file: {e}")
        sys.exit(1)
    
    coordinator = Coordinator(port, persistence_time)
    coordinator.start()

if __name__ == "__main__":
    main()

File: Project3/Coordinator.java
import java.io.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class Coordinator {
    private final int port;
    private final int persistenceTime; // Temporally-bound persistence threshold
    private final Map<String, ParticipantInfo> participants = new ConcurrentHashMap<>();
    private final List<Message> messages = Collections.synchronizedList(new ArrayList<>());

    public Coordinator(int port, int persistenceTime) {
        this.port = port;
        this.persistenceTime = persistenceTime;
    }

    public void start() {
        try (ServerSocket serverSocket = new ServerSocket(port)) {
            //serverSocket.setReuseAddress(true);                         // allow fast restart
            //serverSocket.bind(new InetSocketAddress(port));             // explicit bind
            System.out.println("Coordinator started on port " + port);
            while (true) {
                Socket clientSocket = serverSocket.accept();
                new Thread(() -> handleClient(clientSocket)).start();
            }
        } catch (IOException e) {
            System.out.println("Error starting Coordinator: " + e.getMessage());
        }
    }

    private void handleClient(Socket clientSocket) {
        try (BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true)) {

            String data = in.readLine();
            if (data != null) {
                System.out.println("Received: " + data);
                processRequest(data);
                out.println("{\"status\": \"ack\"}");
            }
        } catch (IOException e) {
            System.out.println("Error handling client: " + e.getMessage());
        }
    }

    private void processRequest(String request) {
        String[] parts = request.split(" ");
        String command = parts[0];
        String participantId;

        // Log the received command
        System.out.println("Received command: " + request);

        switch (command) {
            case "register":
                if (parts.length != 4) {
                    System.out.println("Error: Invalid register format.");
                    return;
                }
                participantId = parts[1];
                String ip = parts[2];
                int threadBPort = Integer.parseInt(parts[3]);
                participants.put(participantId, new ParticipantInfo(participantId, ip, threadBPort, "online"));
                System.out.println("Participant " + participantId + " registered at IP " + ip + " and port " + threadBPort);
                break;

            case "deregister":
                if (parts.length != 2) {
                    System.out.println("Error: Invalid deregister format.");
                    return;
                }
                participantId = parts[1];
                participants.remove(participantId);
                System.out.println("Participant " + participantId + " deregistered.");
                break;

            case "disconnect":
                if (parts.length != 2) {
                    System.out.println("Error: Invalid disconnect format.");
                    return;
                }
                participantId = parts[1];
                ParticipantInfo p1 = participants.get(participantId);
                if (p1 != null) {
                    p1.setStatus("offline");
                    System.out.println("Participant " + participantId + " disconnected.");
                } else {
                    System.out.println("Error: Participant " + participantId + " not found.");
                }
                break;

            case "reconnect":
                if (parts.length != 4) {
                    System.out.println("Error: Invalid reconnect format.");
                    return;
                }
                participantId = parts[1];
                //int newPort = Integer.parseInt(parts[2]);
                int newPort = Integer.parseInt(parts[3]);
                ParticipantInfo p2 = participants.get(participantId);
                if (p2 != null) {
                    p2.setPort(newPort);
                    p2.setStatus("online");
                    System.out.println("Participant " + participantId + " reconnected on port " + newPort);
                    sendPendingMessages(p2);
                } else {
                    System.out.println("Error: Participant " + participantId + " not found.");
                }
                break;

            case "msend":
                if (parts.length < 3) {
                    System.out.println("Error: Invalid msend format.");
                    return;
                }
                participantId = parts[1];
                String message = String.join(" ", Arrays.copyOfRange(parts, 2, parts.length));
                //System.out.println("Message from " + participantId + ": " + message);

                Message m = new Message(participantId, message, System.currentTimeMillis());
                messages.add(m);

                // immediate fan‑out to all **online** members
                sendMessagesToOnlineParticipants(m);
                System.out.println("Multicast from " + participantId + ": " + message);
                break;

            default:
                System.out.println("Error: Unknown command - " + command);
        }
    }

    private void sendPendingMessages(ParticipantInfo pi) {
        long horizon = System.currentTimeMillis() - (persistenceTime * 1000L);
        for (Message msg : messages) {
            if (msg.getTimestamp() <= horizon) continue;                 // outside td window
            if (msg.getTimestamp() <= pi.getLastDelivered()) continue;   // duplicate
            sendSingleMessage(pi, msg);
        }
    }

    private void sendMessagesToOnlineParticipants(Message fresh) {
        for (ParticipantInfo pi : participants.values()) {
            if (!"online".equals(pi.getStatus())) 
                continue;
            sendSingleMessage(pi, fresh);
        }
    }

    private void sendSingleMessage(ParticipantInfo pi, Message msg) {
        try (Socket s = new Socket(pi.getIp(), pi.getPort());
             PrintWriter out = new PrintWriter(s.getOutputStream(), true)) 
             {
                out.println("msend " + msg.getSenderId() + " " + msg.getMessage());
                pi.setLastDelivered(msg.getTimestamp());          // mark delivered
                System.out.printf("[TX] \"%s\" -> %s%n", msg.getMessage(), pi.getId());
    
        } catch (IOException e) {
            System.out.println("Error sending to " + pi.getId() + ": " + e.getMessage());
            pi.setStatus("offline");                          // mark unreachable
        }
    }

    private void sendMessagesToParticipant(String participantId) {
        ParticipantInfo participant = participants.get(participantId);
        if (participant == null || !participant.getStatus().equals("online"))
            return;

        long currentTime = System.currentTimeMillis();
        long thresholdTime = currentTime - (persistenceTime * 1000);

        List<Message> eligibleMessages = messages.stream()
                .filter(msg -> msg.getTimestamp() > thresholdTime)
                .collect(Collectors.toList());

        for (Message message : eligibleMessages) {
            try (Socket socket = new Socket(participant.getIp(), participant.getPort());
                    PrintWriter out = new PrintWriter(socket.getOutputStream(), true)) {

                out.println("msend " + message.getSenderId() + " " + message.getMessage());

            } catch (IOException e) {
                System.out.println("Error sending message to " + participantId + ": " + e.getMessage());
            }
        }
    }

    private void sendMessagesToOnlineParticipants() {
        for (ParticipantInfo participant : participants.values()) {
            if (participant.getStatus().equals("online")) {
                sendMessagesToParticipant(participant.getId());
            }
        }
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java Coordinator <config_file>");
            System.exit(1);
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(args[0]))) {
            int port = Integer.parseInt(reader.readLine().trim());
            int persistenceTime = Integer.parseInt(reader.readLine().trim());
            new Coordinator(port, persistenceTime).start();
        } catch (IOException e) {
            System.out.println("Error reading config file: " + e.getMessage());
        }
    }

    private void broadcastSystemMessage(String message) {
        for (ParticipantInfo participant : participants.values()) {
            if ("online".equals(participant.getStatus())) {
                try (Socket socket = new Socket(participant.getIp(), participant.getPort());
                        PrintWriter out = new PrintWriter(socket.getOutputStream(), true)) {
                    out.println(message);
                } catch (IOException e) {
                    System.out
                            .println("Error sending system message to " + participant.getId() + ": " + e.getMessage());
                }
            }
        }
    }

}
